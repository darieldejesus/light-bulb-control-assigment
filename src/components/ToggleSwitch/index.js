import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const SwitchContainer = styled.div`
  border: 1px solid ${({ active }) => active ? '#59A42B' : '#000'};
  width: 38px;
  height: 24px;
  border-radius: 13px;
  cursor: pointer;
  display: inline-block;
  display: flex;
  flex-direction: column;
  justify-content: center;
  background: ${({ active }) => active ? '#59A42B' : 'unset'};
`;

const SwitchToggle = styled.div`
  border: 1px solid ${({ active }) => active ? '#59A42B' : '#000'};
  width: 22px;
  height: 22px;
  left: ${({ active }) => active ? '14px' : '0'};
  border-radius: 50%;
  background: white;
  position: relative;
  transition: left .2s ease-in-out;
`;

/**
 * Toggle Switch.
 *
 * @param {Object} obj - Main parameter.
 * @param {boolean} obj.active - Status of the toggle button.
 * @param {Function} obj.onClick - Function to handle the onClick event.
 * @returns {Object} ReactElement object.
 * */
function Switch({ active, onClick }) {
  const switchProps = {
    role: 'checkbox',
    tabIndex: 0,
    active,
    onKeyPress: e => onClick(e),
    onClick: e => onClick(e),
    'aria-checked': active,
  };
  return (
    <SwitchContainer {...switchProps}>
      <SwitchToggle active={active} />
    </SwitchContainer>
  );
}

Switch.propTypes = {
  active: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Switch;
