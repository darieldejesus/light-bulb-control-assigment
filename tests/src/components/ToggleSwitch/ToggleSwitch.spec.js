import React from 'react';
import 'jest-styled-components';

import ToggleSwitch from '../../../../src/components/ToggleSwitch';

describe('ToggleSwitch', () => {
  test('Should render ToggleSwitch as inactive', () => {
    const status = false;
    const wrapper = shallow(<ToggleSwitch active={status} onClick={() => {}} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('background', 'unset');
    expect(wrapper.find({ active: status }).at(1)).toHaveStyleRule('left', '0');
  });

  test('Should render ToggleSwitch as active and green', () => {
    const status = true;
    const wrapper = shallow(<ToggleSwitch active={status} onClick={() => {}} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('background', '#59A42B');
    expect(wrapper.find({ active: true }).at(1)).toHaveStyleRule('left', '14px');
  });

  test('Should trigger onClick event after clicking it', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<ToggleSwitch active={false} onClick={onClick} />);
    wrapper.simulate('click');
    expect(onClick).toBeCalled();
  });

  test('Should trigger onKeyPress event', () => {
    const onKeyPress = jest.fn();
    const wrapper = shallow(<ToggleSwitch active={false} onClick={onKeyPress} />);
    wrapper.simulate('keypress', { key: 'Enter' });
  });
});
