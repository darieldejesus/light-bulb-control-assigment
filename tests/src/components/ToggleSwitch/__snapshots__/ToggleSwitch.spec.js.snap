// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch Should render ToggleSwitch as active and green 1`] = `
<styled.div
  active={true}
  aria-checked={true}
  onClick={[Function]}
  onKeyPress={[Function]}
  role="checkbox"
  tabIndex={0}
>
  <styled.div
    active={true}
  />
</styled.div>
`;

exports[`ToggleSwitch Should render ToggleSwitch as inactive 1`] = `
<styled.div
  active={false}
  aria-checked={false}
  onClick={[Function]}
  onKeyPress={[Function]}
  role="checkbox"
  tabIndex={0}
>
  <styled.div
    active={false}
  />
</styled.div>
`;
